src/goto-checker/goto_checker.h

```diff
@@ -0,0 +1,70 @@
+/*******************************************************************\
+
+Module: Goto Checker Interface
+
+Author: Daniel Kroening, Peter Schrammel
+
+\*******************************************************************/
+
+/// \file
+/// Goto Checker Interface
+
+#ifndef CPROVER_GOTO_CHECKER_GOTO_CHECKER_H
+#define CPROVER_GOTO_CHECKER_GOTO_CHECKER_H
+
+#include <goto-programs/goto_trace.h>
+
+#include <util/options.h>
+#include <util/ui_message.h>
+
+#include "properties.h"
+
+/// An implementation of `goto_checkert` provides functionality for
+/// checking a set of properties and returning counterexamples
+/// one by one to the caller.
+/// An implementation of `goto_checkert` is responsible for maintaining
+/// the state of the analysis that it performs (e.g. goto-symex, solver, etc).
+/// It is not responsible for maintaining outcomes (e.g. property results,
+/// counterexamples, etc). However, an implementation may restrict the
+/// sets of properties it is asked to check (e.g. only sequences of subsets).
+class goto_checkert
+{
+public:
+  enum class statust
+  {
+    /// The goto checker may be able to determine the results for more
+    /// properties if operator() is called again.
+    HAVE_MORE,
+    /// The goto checker has returned all results for the given set
+    /// of properties.
+    DONE
+  };
+
+  /// Check whether the given properties with status UNKNOWN or newly
+  /// discovered properties hold
+  /// \param [out] properties: Updates those properties whose
+  ///   results have been determined. Newly discovered properties are added.
+  /// \return whether the goto checker may HAVE_MORE results or
+  ///   whether it is DONE with the given properties.
+  /// After returning DONE, another call to operator() with the same
+  /// properties will return DONE and leave the properties' results unchanged.
+  /// If there is a property with status FAIL then its counterexample
+  /// can be retrieved by calling `build_error_trace` before any
```

Comment:
This sounds like it binds an implementation into an odd stateful implementation, how about just passing a boolean to `operator()` to indicate whether you want traces built or not?

Reply:
1. That will not make it less stateful if you still have to call a function to retrieve the trace.
2. A bool `should_build_trace` together with `optional<goto_tracet>` has the problem of what to return when not buildlng a trace and the status is HAVE_MORE... an empty trace?
2. One could have two operators: one that returns `optional<goto_tracet>` and another one that returns a status. The caller can pick which one to use depending on whether they want traces or not.

Reply:
I was envisaging something like `std::pair<statust, optionalt<tracet>>`, but option 3 sounds pretty good to me

Reply:
Hm... having to decide whether I want a trace or not before actually knowing which property failed is a limitation for certain algorithms (e.g. synthesising invariants while checking properties: I don't need a counterexample when I fail due to a non-inductive invariant candidate, but I'd like to have the counterexample when a property fails). 
So, I think the two operations should be separate. I'm not worried about the statefulness because it is quite natural.